bam_sort.c

```diff
@@ -2012,18 +2064,215 @@ static int ks_radixsort(size_t n, bam1_tag *buf, const sam_hdr_t *h)
     return ret;
 }
 
+/*
+ * Computes the minhash of a sequence using both forward and reverse strands.
+ *
+ * This is used as a sort key for unmapped data, to collate like sequences
+ * together and to improve compression ratio.
+ *
+ * The minhash is returned and *pos filled out with location of this hash
+ * key in the sequence if pos != NULL.
+ */
+static uint64_t minhash(bam1_t *b, int kmer, int *pos, int *rev) {
+    uint64_t hashf = 0, minhashf = UINT64_MAX;
+    uint64_t hashr = 0, minhashr = UINT64_MAX;
+    int minhashpf = 0, minhashpr = 0, i;
+    uint64_t mask = (1L<<(2*kmer))-1;
+    unsigned char *seq = bam_get_seq(b);
+    int len = b->core.l_qseq;
+
+    // Lookup tables for bam_seqi to 0123 fwd/rev hashes
+    // =ACM GRSV TWYH KDBN
+#define X 0
+    unsigned char L[16] = {
+        X,0,1,X,  2,X,X,X,  3,X,X,X,  X,X,X,X,
+    };
+    uint64_t R[16] = {
+        X,3,2,X,  1,X,X,X,  0,X,X,X,  X,X,X,X,
+    };
+    for (i = 0; i < 16; i++)
+        R[i] <<= 2*(kmer-1);
+
+    // Punt homopolymers somewhere central in the hash space
+#define XOR (0xdead7878beef7878 & mask)
+
+    // Initialise hash keys
+    for (i = 0; i < kmer-1 && i < len; i++) {
+        int base = bam_seqi(seq, i);
+        hashf = (hashf<<2) | L[base];
+        hashr = (hashr>>2) | R[base];
+    }
+
+    // Loop to find minimum
+    for (; i < len; i++) {
+        int base = bam_seqi(seq, i);
+
+        hashf = ((hashf<<2) | L[base]) & mask;
+        hashr =  (hashr>>2) | R[base];
+
+        if (minhashf > (hashf^XOR))
+            minhashf = (hashf^XOR), minhashpf = i;
+        if (minhashr > (hashr^XOR))
+            minhashr = (hashr^XOR), minhashpr = len-i+kmer-2;
+
+    }
+
+    if (minhashf <= minhashr) {
+        if (rev) *rev = 0;
+        if (pos) *pos = minhashpf;
+        return minhashf;
+    } else {
+        if (rev) *rev = 1;
+        if (pos) *pos = minhashpr;
+        return minhashr;
+    }
+}
+
+//--- Start of candidates to punt to htslib
+/*!
+ * @abstract
+ * Extracts the sequence (in current alignment orientation) from
+ * a bam record and places it in buf, which is nul terminated.
+ *
+ * @param b     The bam structure
+ * @param buf   A buffer at least b->core.l_qseq+1 bytes long
+ */
+void bam_to_seq(bam1_t *b, char *buf) {
```

Comment:
Should be `static void ...` (also other functions below).

Reply:
Done.

In a second commit for sake of consistency I've updated all the other samtools functions (of which there are *MANY*) given this is now our new coding style.  However it may cause problems for people who manipulate internals of samtools via their own nefarious means.  Can't have it both ways!